feat(similaritySearch): adding similaritySearch methods

Itzaprado opened the pull request:
[ ] similaritySearch
[ ] similaritySearchByVector
[ ] similaritySearchWithScore
[ ] similaritySearchVectorWithScore
[ ] queryCollection

Itzaprado committed:
feat(similaritySearch): adding similaritySearch methods
[ ] similaritySearch
[ ] similaritySearchByVector
[ ] similaritySearchWithScore
[ ] similaritySearchVectorWithScore
[ ] queryCollection

Itzaprado committed:
fix: making queryCollection private

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -293,6 +290,68 @@ class PostgresVectorStore extends VectorStore {
+   * @param _callbacks - Optional callbacks for monitoring search progress
+   * @returns A promise resolving to an array of `DocumentInterface` instances representing similar documents.
+   */
+  async similaritySearch(query: string, k: number = 4, filter?: this["FilterType"] | undefined, _callbacks?: Callbacks | undefined): Promise<Document[]> {
```
We can remove this since the interface will inject this method

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -293,6 +290,68 @@ class PostgresVectorStore extends VectorStore {
+    return await this.similaritySearchByVector(embedding, k, filter)
+  }
+
+  async similaritySearchByVector(embedding: number[], k?: number, filter?: this["FilterType"]): Promise<Document[]> {
```
We won't include this for now since it's not needed by the interface

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -293,6 +290,68 @@ class PostgresVectorStore extends VectorStore {
+    return documents;
+  }
+
+  async similaritySearchVectorWithScore(embedding: number[], k?: number, filter?: this["FilterType"]): Promise<[Document, number][]> {
```
k is required

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -293,6 +290,68 @@ class PostgresVectorStore extends VectorStore {
+    const _filter = filter !== undefined ? `WHERE "${Object.entries(filter)[0][0]}" = '${Object.entries(filter)[0][1]}'` : "";
+    let results;
+
+    const query = `SELECT *, ${searchFunction}("${this.embeddingColumn}", '[${embedding}]') as distance FROM "${this.schemaName}"."${this.tableName}" ${_filter} ORDER BY "${this.embeddingColumn}" ${operator} '[${embedding}]' LIMIT ${k};` 
```
Instead of "*" (selecting all), we should only select the columns we need ie the id, content, metadata columns, and the metadata json column. They all need to have quotes around the column names too.

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -293,6 +290,68 @@ class PostgresVectorStore extends VectorStore {
+
+    if (this.indexQueryOptions) {
+      results = await this.engine.pool.raw(`SET LOCAL ${this.indexQueryOptions.to_string()}`) // Add the toString method to QueryOptions
+    } else {
```
remove else. The query should always execute

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/vectorStore.test.ts:
```diff
@@ -209,12 +210,48 @@ describe("VectorStore methods", () => {
 
-  afterEach(async () => {
-    await PEInstance.pool.raw(`TRUNCATE TABLE "${CUSTOM_TABLE}"`);
+  test("similaritySearch", async () => {
```
Can you add some tests with filters?

Itzaprado replied:
Could you please specify what filters you would like me to add? I added one on line 217, is that what you are referring to? or would you like it to be in a separate test?

averikitsch replied:
Sorry, I missed this. Can we update it to use the metadata like ``"page" = '2'``?

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -293,6 +290,68 @@ class PostgresVectorStore extends VectorStore {
+    k = k ?? this.k;
+    const operator = this.distanceStrategy.operator;
+    const searchFunction = this.distanceStrategy.searchFunction;
+    const _filter = filter !== undefined ? `WHERE "${Object.entries(filter)[0][0]}" = '${Object.entries(filter)[0][1]}'` : "";
```
To simplify filters, let's update to the this.FilterType to just a string
```suggestion
    const _filter = filter !== undefined ? `WHERE ${filter}` : "";
```

Itzaprado committed:
refactor: remove unnecesary methods
[ ] change FilterType to string type
[ ] update queryCollection method
[ ] update and remove tests

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/vectorStore.test.ts:
```diff
@@ -209,12 +210,39 @@ describe("VectorStore methods", () => {
+    const results = await vectorStoreInstance.similaritySearch("foo", 1);
+    expect(results.length).toBe(1)
+
+    const results_2 = await vectorStoreInstance.similaritySearch("foo", 1,  `"my_content" = 'bar'`)
```
```suggestion
    const results_2 = await vectorStoreInstance.similaritySearch("foo", 1,  `"page" = '2'`)
```

averikitsch committed:
Update libs/langchain-google-cloud-sql-pg/src/tests/vectorStore.test.ts
